src/admin/metrics.rs

```diff
@@ -0,0 +1,336 @@
+    let mut interval = tokio::time::interval(Duration::from_millis(10));
+    loop {
+        interval.tick().await;
+        while let Some(incoming) = metrics_rx.recv().await {
+            incoming.requests_complete("test", "test", &200, Duration::from_millis(100));
+            let test_report = metrics_encoder(registry_state.clone()).await;
+            log_info!("prometheus metrics: {:?}", test_report);
+            let _registry = registry_state.read().unwrap();
+            // incoming.requests_complete(path, method, status, duration);
+        }
+    }
+}
+/// Accepts metrics request, encodes and prints
+#[cfg(feature = "prometheusd")]
+pub async fn accept_metrics_request(
+    tx: Request<hyper::body::Incoming>,
+    metrics_tx: Arc<RwLock<MetricSender>>,
+    registry_state: Arc<RwLock<StorageRegistry>>,
+) -> Result<hyper::Response<Full<Bytes>>, Infallible> {
+    // if tx.uri().path() == "/ws_metrics" {
+    //     return accept_ws_metrics(metrics_tx).await;
+    // } else if tx.uri().path() == "/http_metrics" {
+    //     return accept_http_metrics(metrics_tx).await;
+    // }
+    use crate::balancer::format::incoming_to_value;
+    use serde_json::{
+        json,
+        Value,
+        Value::Null,
+    };
+    // let mut tx = incoming_to_value(tx).await;
+    //     tx = json!({
+    //         "path": tx["path"],
+    //         "method": tx["method"],
+    //         "status": tx["status"],
+    //         "duration": tx["duration"],
+    //     });
+
+    let metrics_report = metrics_encoder(registry_state).await;
+    let response = Ok(hyper::Response::builder()
+        .status(200)
+        .body(Full::from(Bytes::from(metrics_report)))
+        .unwrap());
+    (response)
+}
+#[cfg(feature = "prometheusd")]
+async fn accept_ws_metrics(
+    _metrics_tx: Arc<MetricSender>,
+) -> Result<hyper::Response<Full<Bytes>>, Infallible> {
+    unimplemented!()
+}
+#[cfg(feature = "prometheusd")]
+async fn accept_http_metrics(
+    _metrics_tx: Arc<MetricSender>,
+) -> Result<hyper::Response<Full<Bytes>>, Infallible> {
+    unimplemented!()
+}
+#[cfg(feature = "prometheusd")]
+async fn metrics_encoder(storage_registry: Arc<RwLock<StorageRegistry>>) -> String {
+    use prometheus::Encoder;
+    let encoder = prometheus::TextEncoder::new();
+    let mut buffer = vec![];
+    let registry = storage_registry.read().unwrap().gather();
+    encoder.encode(&registry, &mut buffer).unwrap();
+    String::from_utf8(buffer).unwrap()
+}
+#[cfg(feature = "prometheusd")]
+pub async fn metrics_monitor(
+    metrics_rx: MetricReceiver,
+    storage_registry: Arc<RwLock<StorageRegistry>>,
+) {
+    //TODO: figure ownership mess here
+    let registry;
+    {
+        let registry_guard = storage_registry.read().unwrap();
+        registry = registry_guard;
+    }
+    let metrics_status = Arc::new(RwLock::new(
+        RpcMetrics::instance(&registry).unwrap().to_owned(),
+    ));
+    let metrics_stat_listener = metrics_status.clone();
```

Comment:
id just not deal with creating a new task and dealing with sendin. have all of this setup and accept loop be in the same thread